### PR TITLE
mgr/DaemonServer.cc: prevent mgr crashes caused by integer underflow that is triggered by large increases to pg_num/pgp_num

### PR DESCRIPTION
**Pull request by cfsnyder**
Signed-off-by: Cory Snyder <csnyder@iland.com>

This fixes a scenario where mgrs continually crash while attempting to apply large increases to pg_num/pgp_num. The max step size (estmax) for each incremental update to the pgp_num is calculated as a percentage of the pg_num, which permits the possibility for the max step size (estmax) to be greater than the current pgp_num when the increase is large; this causes an integer underflow when the max step size is subtracted from the pgp_num in order to calculate the next step size with std::clamp. The integer underflow causes hi < lo in args passed to std::clamp, which causes a failed assertion, SIGABRT, and ultimately crashing mgr.

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -2831,8 +2831,12 @@ void DaemonServer::adjust_pgs()
 				 max_misplaced / 2.0);
 	      unsigned estmax = std::max<unsigned>(
 		(double)p.get_pg_num() * room, 1u);
+	      unsigned nextmin = 0;
```
Comment (tchaikov) [inline src/mgr/DaemonServer.cc:2834]:
```suggestion
	      unsigned next_min = 0;
```

**Review by tchaikov — APPROVED** [review]
just a nit.

agreed. it's unusual. but possible and actually happens..

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (cfsnyder) [reply src/mgr/DaemonServer.cc]: agreed, this is better! just updated

**Comment by cfsnyder** [comment]
Note that the bug is quite easy to reproduce. It will occur systematically when updating pgp_num to track a a recent increase to pg_num if the following condition is true:
`target_max_misplaced_ratio > (2 * pgp_num / pg_num)`.

For example, with the default `target_max_misplaced_ratio` of 0.05 and starting with 32 PGs, it will always manifest itself when scaling to >= 2048 PGs.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-03_03:42:24-rados-wip-kefu-testing-2021-06-03-1002-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-04_02:22:09-rados-wip-kefu-testing-2021-06-03-2320-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-04_08:38:24-rados-wip-kefu-testing-2021-06-04-1522-distro-basic-smithi/

**Pull request merged**